### DISCUSSION TITLE
Support dynamic "origins" for this library for testing

### DISCUSSION THREAD

**Discussion by witwaycorp** [passwordless-lib/fido2-net-lib#538] [General]
All,

I have a WebAuthn server using this library and was hoping to modify it so that I can support any sub-domain for the main website for some test scenarios I am working on. Example, the app is deployed to auth.example.com. I however would like to support any sub-domain of example.com like dev.example.com, prod.example.com, etc. 

many thanks!

**Comment by abergs** [comment]
Not sure I follow - you can initiate a new instance of the library with the appropriate configuration for your origin.

**Reply by witwaycorp** [reply]
Yes I have a custom origin deployed but want to be able to use any or a list of sub-domains vs just the two that are typically supported like auth.example.com and example.com.

**Reply by abergs** [reply]
Are you referring to the proposed new feature of multi-origins in the WebAuthn L3 spec?

**Reply by witwaycorp** [reply]
No. Just being able to ensure that the origin website is indeed part of the parent domain by checking the origin like this:
if (origin == rp.id || origin.Contains("." + rp.id))
{
    // origin domain is valid. Continue processing.
}

**Reply by jackaboy7981** [reply]
Hey is there a option to set the origin configuration of fido2 to allow all subdomains of a particular domain?

becuse i have a scenario where i will be supporting new subdomains while the project is running so need to add the origins dynamically or set the origin such that it accepts all subdomains

if there is no option available currently ,  is there any work around for this?

**Reply by abergs** [reply]
@jackaboy7981 The suggested workaround is to instantiate a Fido2-class with configuration through code, there you can perform any type of dynamic/multi-origin support, pseudo code example:

```csharp

var origin = new Uri(incomingRequest.Request.Host); // use domain from the incoming request based on headers etc.
builder.Services.AddFido2(options =>
{
    options.ServerDomain = origin.Host;
    options.ServerName = "FIDO2 Server";
    options.Origins = new HashSet<string> { origin.AbsoluteUri };
    options.TimestampDriftTolerance = 1000;
});

```

**Reply by jackaboy7981** [reply]
This code should be part of startup right ?
If not, won't this cause service re-registration issues ?

In my use case i will be handling n-number of subdomains and the request should only be accepted from my subdomains.

**Reply by abergs** [reply]
Not using DI - or using DI with a "scoped" factory like above would match your requirements.

**Reply by Regenhardt** [reply]
For usage with DI, you can inject something like this [FidoFactory](https://gitlab.com/Regenhardt/fido2-webauthn-blazor/-/blob/master/Server/FidoFactory.cs) and use your own URL-matching logic to decide which hosts you want to allow. Example usage in this [UserController](https://gitlab.com/Regenhardt/fido2-webauthn-blazor/-/blob/master/Server/Controllers/UserController.cs?ref_type=heads#L79).

**Reply by HakooDesai** [reply]
According to spec, if you set RP ID as "example.com" during registration, it should support subdomains as login.example.com" and "auth.example.com".  @abergs , Can you please confirm this?

**Reply by abergs** [reply]
@HakooDesai yes that's a valid behavior. The client (e.g. browser) will refuse to show passkeys outside of the rpid (only create/show keys for the current domain)

However - the RP server library may have their own logic for creating passkeys on multiple domains, e.g. multi tenant servers. Users would
Only be able to exercise the passkey on the domain it was created for.

I'm using domain freely here, as in domain + subdomains.
